File: 0109_Convert_Sorted_List_to_Binary_Search_Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        return toBST(head, NULL);
    }
    TreeNode* toBST(ListNode* head, ListNode* tail) {
        if (head == tail) return NULL;
        if (head->next == tail) {
            TreeNode* tRoot = new TreeNode(head->val);
            return tRoot;
        }
        ListNode* fast = head;
        ListNode* slow = head;
        while (fast != tail && fast->next != tail) {
            fast = fast->next->next;
            slow = slow->next;
        }
        TreeNode* tRoot = new TreeNode(slow->val);
        tRoot->left = toBST(head, slow);
        tRoot->right = toBST(slow->next, tail);
        return tRoot;
    }
};

File: 0109_Convert_Sorted_List_to_Binary_Search_Tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def getLen(self, head):
        size = 0
        while (head):
            size += 1
            head = head.next
        return size
    
    def generate(self, n):
        if n == 0:
            return None
        root = TreeNode(0)
        root.left = self.generate(n // 2)
        root.val = self.node.val
        self.node = self.node.next
        root.right = self.generate(n - (n // 2) - 1)
        return root

    def sortedListToBST(self, head: ListNode) -> TreeNode:
        self.node = head
        return self.generate(self.getLen(head))


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMid(self, head):
        slow = fast = head
        pre = slow
        while fast and fast.next:
            pre = slow
            slow = slow.next
            fast = fast.next.next

        return pre, slow

    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        if not head:
            return None
        pre, mid = self.findMid(head)
        root = TreeNode(mid.val)
        pre.next = None
        if mid != head:
            root.left = self.sortedListToBST(head)
        root.right = self.sortedListToBST(mid.next)
        return root
